In [1]:
import os
import sys
sys.path.append('..')


In [2]:
# Cell 1: Imports
from src.model_training import run_credit_pipeline, run_fraud_pipeline

# Cell 2: Run for Credit Card Data
run_credit_pipeline()

# Cell 3: Run for Fraud (E-commerce) Data
run_fraud_pipeline()


c:\Users\tsega\OneDrive\Documents\-detection-of-fraud-cases-for-e-commerce\.venv\Lib\site-packages\xgboost\training.py:183: UserWarning: [17:18:46] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)



--- CreditCard | LogisticRegression ---
[[55397  1467]
 [    8    90]]
              precision    recall  f1-score   support

           0       1.00      0.97      0.99     56864
           1       0.06      0.92      0.11        98

    accuracy                           0.97     56962
   macro avg       0.53      0.95      0.55     56962
weighted avg       1.00      0.97      0.99     56962

ROC AUC: 0.9708434302252134
PR AUC : 0.724469435669471

--- CreditCard | XGBClassifier ---
[[56833    31]
 [   15    83]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     56864
           1       0.73      0.85      0.78        98

    accuracy                           1.00     56962
   macro avg       0.86      0.92      0.89     56962
weighted avg       1.00      1.00      1.00     56962

ROC AUC: 0.9799832109264639
PR AUC : 0.8682821332846721


c:\Users\tsega\OneDrive\Documents\-detection-of-fraud-cases-for-e-commerce\.venv\Lib\site-packages\xgboost\training.py:183: UserWarning: [17:18:52] WARNING: C:\actions-runner\_work\xgboost\xgboost\src\learner.cc:738: 
Parameters: { "use_label_encoder" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)



--- E-Commerce Fraud | LogisticRegression ---
[[17588  9805]
 [  864  1966]]
              precision    recall  f1-score   support

           0       0.95      0.64      0.77     27393
           1       0.17      0.69      0.27      2830

    accuracy                           0.65     30223
   macro avg       0.56      0.67      0.52     30223
weighted avg       0.88      0.65      0.72     30223

ROC AUC: 0.7518792490253436
PR AUC : 0.45181107105591145

--- E-Commerce Fraud | XGBClassifier ---
[[27313    80]
 [ 1333  1497]]
              precision    recall  f1-score   support

           0       0.95      1.00      0.97     27393
           1       0.95      0.53      0.68      2830

    accuracy                           0.95     30223
   macro avg       0.95      0.76      0.83     30223
weighted avg       0.95      0.95      0.95     30223

ROC AUC: 0.7642964872380412
PR AUC : 0.6057035996545085


 Model Performance Summary and Interpretation
 Dataset 1: Credit Card Fraud Detection
Model	Accuracy	Precision (Fraud)	Recall (Fraud)	F1-Score (Fraud)	ROC AUC	PR AUC
Logistic Regression	0.97	0.06	0.92	0.11	0.971	0.724
XGBoost Classifier	1.00	0.73	0.85	0.78	0.980	0.868

Key Observations:

The baseline Logistic Regression achieves very high recall (92%) but extremely poor precision (6%), indicating many false positives.

XGBoost significantly outperforms LR, achieving strong precision–recall balance (F1 = 0.78), and higher ROC AUC and PR AUC scores.

PR AUC is especially important here due to high class imbalance, and XGBoost's 0.868 shows strong performance in detecting rare fraud cases.

 Dataset 2: E-Commerce Fraud Detection
Model	Accuracy	Precision (Fraud)	Recall (Fraud)	F1-Score (Fraud)	ROC AUC	PR AUC
Logistic Regression	0.65	0.17	0.69	0.27	0.752	0.452
XGBoost Classifier	0.95	0.95	0.53	0.68	0.764	0.606

Key Observations:

Logistic Regression again shows high recall but very low precision (many false alarms).

XGBoost achieves strong overall performance — very high precision (95%) but slightly lower recall (53%).

F1-score and PR AUC are better with XGBoost, making it the better choice.

##  Which Model is Best?

| **Criterion**              | **Winner**        | **Why**                                                                 |
|---------------------------|-------------------|-------------------------------------------------------------------------|
| Overall Accuracy           | XGBoost           | Near-perfect accuracy on both datasets                                  |
| Fraud Detection (Recall)   | Logistic (CC), XGB (EC) | Logistic achieves 92% recall on credit card, but XGB balances both metrics better in e-commerce |
| PR AUC (Best for Imbalance) | XGBoost           | Handles minority class well; better PR AUC in both cases                |
| Balanced F1-Score          | XGBoost           | Stronger fraud classification with fewer false positives                |


 Conclusion
Across both fraud datasets, XGBoost consistently outperforms Logistic Regression, especially in balancing precision and recall on highly imbalanced data. It is the preferred model for deployment in both e-commerce and credit card fraud detection tasks.